pkg/kafka/correlation_cache.go

```diff
@@ -0,0 +1,224 @@
+		cache:              requestsCache{},
+		nextSequenceNumber: 1,
+		stopGc:             make(chan struct{}),
+	}
+
+	go cc.garbageCollector()
+
+	return cc
+}
+
+// DeleteCache releases the cache and stops the garbage collector. This
+// function must be called when the cache is no longer required, otherwise go
+// routines are leaked.
+func (cc *CorrelationCache) DeleteCache() {
+	close(cc.stopGc)
+}
+
+// HandleRequest must be called when a request is forwarded to the broker, will
+// keep track of the request and rewrite the correlation ID inside of the
+// request to a sequence number. This sequence number is guaranteed to be
+// unique within the connection covered by the cache.
+func (cc *CorrelationCache) HandleRequest(req *RequestMessage, finishFunc FinishFunc) {
+	cc.mutex.Lock()
+	defer cc.mutex.Unlock()
+
+	// save the original correlation ID
+	origCorrelationID := req.GetCorrelationID()
+
+	// Use a sequence number to generate a correlation ID that is
+	// guaranteed to be unique
+	newCorrelationID := cc.nextSequenceNumber
+	cc.nextSequenceNumber++
+
+	// Overwrite the correlation ID in the request to allow correlating the
+	// response later on. The original correlation ID will be restored when
+	// forwarding the response
+	req.SetCorrelationID(newCorrelationID)
+
+	if _, ok := cc.cache[newCorrelationID]; ok {
+		log.Warning("BUG: Overwriting Kafka request message in correlation cache")
+	}
+
+	cc.cache[newCorrelationID] = &correlationEntry{
+		request:           req,
+		created:           time.Now(),
+		origCorrelationID: origCorrelationID,
+		finishFunc:        finishFunc,
+	}
+}
+
+// Delete removes a request from the correlation cache. After deletion, the
+// request will no longer be correlated with responses. Note: Calling Delete()
+// will *not* invoke the FinishFunc of the request
+func (cc *CorrelationCache) Delete(req *RequestMessage) {
+	cc.mutex.Lock()
+	defer cc.mutex.Unlock()
+
+	correlationID := req.GetCorrelationID()
+	if correlationID != 0 {
+		delete(cc.cache, correlationID)
+	}
+}
+
+// correlate returns the request message with the matching correlation ID
+func (cc *CorrelationCache) correlate(id CorrelationID) *correlationEntry {
+	cc.mutex.RLock()
+	defer cc.mutex.RUnlock()
+
+	entry, _ := cc.cache[id]
+	return entry
+}
+
+// CorrelateResponse extracts the correlation ID from the response message,
+// correlates the corresponding request, restores the original correlation ID
+// in the response and returns the original request
+func (cc *CorrelationCache) CorrelateResponse(res *ResponseMessage) *RequestMessage {
+	entry := cc.correlate(res.GetCorrelationID())
+	if entry != nil {
+
+		res.SetCorrelationID(entry.origCorrelationID)
+
+		if entry.finishFunc != nil {
+			entry.finishFunc(entry.request)
+		}
+
+		cc.Delete(entry.request)
+		return entry.request
+	}
+
+	return nil
+}
+
+func (cc *CorrelationCache) garbageCollector() {
+	for {
+		select {
+		case <-cc.stopGc:
+			return
+		default:
```

Comment:
This might be a `go specific thing` which I don't understand, but why is default needed here?

Reply:
Writing it this way makes it the most obvious to me that we are checking if the channel has been closed without blocking.